{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from glob import glob\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm import tqdm\n",
    "import torch\n",
    "from torch_geometric.data import Data\n",
    "from torch_geometric.data import DataLoader\n",
    "\n",
    "import pickle\n",
    "import sys\n",
    "import requests\n",
    "\n",
    "from types import MethodType\n",
    "import importlib\n",
    "from scperturb import *\n",
    "\n",
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.18.0'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "import celloracle as co\n",
    "co.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import roc_curve, auc, precision_recall_curve\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append(\"/data1/lichen/code/single_cell_perturbation/scPerturb/Byte_Pert_Data/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import v1\n",
    "from v1.utils import *\n",
    "from v1.dataloader import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'v1.dataloader' from '/data1/lichen/code/single_cell_perturbation/scPerturb/Byte_Pert_Data/v1/dataloader.py'>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "importlib.reload(v1)\n",
    "importlib.reload(v1.utils)\n",
    "importlib.reload(v1.dataloader)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# dataset metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pert_dict = \\\n",
    "{'AdamsonWeissman2016_GSM2406675_10X001': {'perturbation_type': 'CRISPRi',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 0},\n",
    " 'AdamsonWeissman2016_GSM2406677_10X005': {'perturbation_type': 'CRISPRi',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 1,\n",
    "  'multi_sgRNA': 0},\n",
    " 'AdamsonWeissman2016_GSM2406681_10X010': {'perturbation_type': 'CRISPRi',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 0},\n",
    "#  'DatlingerBock2017': {'perturbation_type': 'CRISPR-cas9',\n",
    "#   'perturbation_state': 1,\n",
    "#   'multi_perturbation': 0,\n",
    "#   'multi_sgRNA': 1},\n",
    " 'DatlingerBock2021': {'perturbation_type': 'CRISPR-cas9',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 1},\n",
    " 'DixitRegev2016': {'perturbation_type': 'CRISPR-cas9',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 1},\n",
    "#  'FrangiehIzar2021_RNA': {'perturbation_type': 'CRISPR-cas9', # 唯一一个multi celltype，先留着\n",
    "#   'perturbation_state': 1,\n",
    "#   'multi_perturbation': 0,\n",
    "#   'multi_sgRNA': 1},\n",
    " \n",
    "#  'LaraAstiasoHuntly2023_exvivo': {'perturbation_type': 'CRISPR-cas9',\n",
    "#   'perturbation_state': 3,\n",
    "#   'multi_perturbation': 0,\n",
    "#   'multi_sgRNA': 1},\n",
    "#  'LaraAstiasoHuntly2023_invivo': {'perturbation_type': 'CRISPR-cas9',\n",
    "#   'perturbation_state': 3,\n",
    "#   'multi_perturbation': 0,\n",
    "#   'multi_sgRNA': 1},\n",
    "#  'LaraAstiasoHuntly2023_leukemia': {'perturbation_type': 'CRISPR-cas9',\n",
    "#   'perturbation_state': 3,\n",
    "#   'multi_perturbation': 0,\n",
    "#   'multi_sgRNA': 1},\n",
    "#  'LiangWang2023': {'perturbation_type': 'CRISPR-cas9',\n",
    "#   'perturbation_state': 3,\n",
    "#   'multi_perturbation': 1,\n",
    "#   'multi_sgRNA': 1},\n",
    "#  'McFarlandTsherniak2020': {'perturbation_type': 'CRISPR-cas9',\n",
    "#   'perturbation_state': 3,\n",
    "#   'multi_perturbation': 0,\n",
    "#   'multi_sgRNA': 0},\n",
    " 'NormanWeissman2019_filtered': {'perturbation_type': 'CRISPRa',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 1,\n",
    "  'multi_sgRNA': 0},\n",
    " 'PapalexiSatija2021_eccite_RNA': {'perturbation_type': 'CRISPR-cas9',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 1},\n",
    " 'PapalexiSatija2021_eccite_arrayed_RNA': {'perturbation_type': 'CRISPR-cas9',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 1},\n",
    " 'ReplogleWeissman2022_K562_essential': {'perturbation_type': 'CRISPRi',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 1},\n",
    "#  'ReplogleWeissman2022_K562_gwps': {'perturbation_type': 'CRISPRi', # pert太多了，先不测试\n",
    "#   'perturbation_state': 1,\n",
    "#   'multi_perturbation': 0,\n",
    "#   'multi_sgRNA': 1},\n",
    " 'ReplogleWeissman2022_rpe1': {'perturbation_type': 'CRISPRi',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 1},\n",
    "#  'SantinhaPlatt2023': {'perturbation_type': 'CRISPR-cas9',\n",
    "#   'perturbation_state': 3,\n",
    "#   'multi_perturbation': 0,\n",
    "#   'multi_sgRNA': 0},\n",
    " 'ShifrutMarson2018': {'perturbation_type': 'CRISPR-cas9',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 1,\n",
    "  'multi_sgRNA': 1},\n",
    " 'SunshineHein2023': {'perturbation_type': 'CRISPR-cas9',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 1,\n",
    "  'multi_sgRNA': 1},\n",
    " 'TianKampmann2019_day7neuron': {'perturbation_type': 'CRISPRi',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 1,\n",
    "  'multi_sgRNA': 1},\n",
    " 'TianKampmann2019_iPSC': {'perturbation_type': 'CRISPRi',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 1,\n",
    "  'multi_sgRNA': 1},\n",
    " 'TianKampmann2021_CRISPRa': {'perturbation_type': 'CRISPRa',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 1},\n",
    " 'TianKampmann2021_CRISPRi': {'perturbation_type': 'CRISPRi',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 1},\n",
    " 'XuCao2023': {'perturbation_type': 'CRISPRi',\n",
    "  'perturbation_state': 1,\n",
    "  'multi_perturbation': 0,\n",
    "  'multi_sgRNA': 1}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prefix_list = list(gene_pert_dict.keys())\n",
    "len(prefix_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 在每个cell line上运行CellOracle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# - get cell line name\n",
    "common_cell_line = \\\n",
    "{   'A549': 'A549',\n",
    "    'HEPG2': 'HepG2',\n",
    "    'HT29': 'HT29',\n",
    "    'MCF7': 'MCF7',\n",
    "    # 'SKBR3': 'SK-BR-3',\n",
    "    'SW480': 'SW480',\n",
    "    'PC3': 'PC3',\n",
    "    'A375': 'A375',\n",
    "} # L1000 cell line : single-cell cell line\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 36720 × 978\n",
       "    obs: 'sig_id', 'pert_id', 'pert_iname', 'pert_type', 'cell_id', 'pert_dose', 'pert_dose_unit', 'pert_idose', 'pert_time', 'pert_time_unit', 'pert_itime', 'distil_id'\n",
       "    var: 'pr_gene_id', 'pr_gene_symbol', 'pr_gene_title', 'pr_is_lm', 'pr_is_bing'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# - read adata_L1000, this is processed data\n",
    "adata_L1000 = sc.read('/nfs/public/lichen/results/single_cell_perturbation/perturbation_benchmark/benchmark_data/L1000/GSE92742/adata_gene_pert.h5ad')\n",
    "adata_L1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================== cell line is A549\n",
      "adata.shape is:  (500, 5155)\n",
      "Loading prebuilt promoter base-GRN. Version: hg19_gimmemotifsv5_fpr2\n",
      "base_GRN.shape:  (37003, 1096)\n",
      "file exists\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "710b54794a0045fcafb0cd4c1adcfc51",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "L1000_total_perts num:  3620\n",
      "common_perts num:  306\n",
      "common var to L1000 data is:  933\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "from sklearn.metrics import precision_recall_curve, auc\n",
    "\n",
    "for cell_line_bulk in list(common_cell_line.keys())[:]:\n",
    "    cell_line_single = common_cell_line[cell_line_bulk]\n",
    "    print('='*20, f'cell line is {cell_line_single}')\n",
    "    \n",
    "    #####################################################\n",
    "    \n",
    "    #===================prepare data\n",
    "    if cell_line_bulk in ['PC3', 'A375']:\n",
    "        save_dir_adata = '/nfs/public/lichen/results/single_cell_perturbation/perturbation_benchmark/benchmark_data/L1000/single_cell_data/SCP542'\n",
    "    else:\n",
    "        save_dir_adata = '/nfs/public/lichen/results/single_cell_perturbation/perturbation_benchmark/benchmark_data/L1000/single_cell_data/CNP0003658'\n",
    "    adata_rna = sc.read(os.path.join(save_dir_adata, cell_line_bulk, f'adata_{cell_line_bulk}.h5ad'))\n",
    "\n",
    "    # - consctrut corr mtx\n",
    "    if not isinstance(adata_rna.X, np.ndarray):\n",
    "        adata_rna.X = adata_rna.X.toarray()\n",
    "    # corr_mtx = np.corrcoef(adata_rna.X.T)\n",
    "    \n",
    "    # - get var_names\n",
    "    var_names = list(adata_rna.var_names)\n",
    "    \n",
    "    # - get common pert\n",
    "    adata_L1000_sub = adata_L1000[adata_L1000.obs['cell_id']==cell_line_bulk]\n",
    "    L1000_total_perts = np.unique(adata_L1000_sub.obs['pert_iname'])\n",
    "    \n",
    "    \n",
    "    n_cells_downsample = 10000\n",
    "    threshold_number = 10000\n",
    "    \n",
    "    ##########################################################\n",
    "    \n",
    "    # - get control adata\n",
    "    adata = adata_rna.copy()\n",
    "    adata.obs['celltype'] = cell_line_bulk\n",
    "    print(f'adata.shape is: ',adata.shape)\n",
    "\n",
    "    # -- get the baseGRN\n",
    "    # Load TF info which was made from mouse cell atlas dataset.\n",
    "    base_GRN = co.data.load_human_promoter_base_GRN()\n",
    "    print('base_GRN.shape: ', base_GRN.shape)\n",
    "\n",
    "    tmp_dir = '/nfs/public/lichen/results/single_cell_perturbation/perturbation_benchmark/L1000'\n",
    "    save_prefix = f'CellOracle/{cell_line_bulk}' # use result of K562 to do the direct transfer\n",
    "    os.makedirs(os.path.join(tmp_dir, save_prefix), exist_ok=True)\n",
    "\n",
    "    save_dir = os.path.join(tmp_dir, save_prefix)\n",
    "    if os.path.exists(os.path.join(save_dir, \"ctrl.celloracle.oracle\")):\n",
    "        print('file exists')\n",
    "        oracle = co.load_hdf5(os.path.join(save_dir, \"ctrl.celloracle.oracle\"))\n",
    "        links = co.load_hdf5(file_path=os.path.join(save_dir, \"ctrl.celloracle.links\"))\n",
    "        \n",
    "    else:\n",
    "\n",
    "        # - start CellOracle process for the whole ctrl\n",
    "\n",
    "        # -- keep raw cont data before log transformation\n",
    "        adata.raw = adata\n",
    "        if not isinstance(adata.raw.X, np.ndarray):\n",
    "            adata.layers[\"raw_count\"] = (np.exp(adata.raw.X.toarray())-1).copy()\n",
    "        else:\n",
    "            adata.layers[\"raw_count\"] = (np.exp(adata.raw.X)-1).copy()\n",
    "            \n",
    "        # -- get umap \n",
    "        sc.pp.scale(adata)\n",
    "        # PCA\n",
    "        sc.tl.pca(adata, svd_solver='arpack', random_state=2022)\n",
    "        # UMAP\n",
    "        sc.pp.neighbors(adata, n_neighbors=4, n_pcs=20, random_state=2022)\n",
    "        sc.tl.umap(adata,random_state=2022)\n",
    "\n",
    "        # -- Random downsampling into 30K cells if the anndata object include more than 30 K cells.\n",
    "        if adata.shape[0] > n_cells_downsample:\n",
    "            # Let's dowmsample into 30K cells\n",
    "            sc.pp.subsample(adata, n_obs=n_cells_downsample, random_state=123)\n",
    "        print(f\"Cell number is :{adata.shape[0]}\")\n",
    "\n",
    "        # -- Instantiate Oracle object\n",
    "        oracle = co.Oracle()\n",
    "\n",
    "        # -- Check data in anndata\n",
    "        print(\"Metadata columns :\", list(adata.obs.columns))\n",
    "        print(\"Dimensional reduction: \", list(adata.obsm.keys()))\n",
    "\n",
    "        # -- In this notebook, we use the unscaled mRNA count for the nput of Oracle object.\n",
    "        adata.X = adata.layers[\"raw_count\"].copy()\n",
    "\n",
    "        # -- Instantiate Oracle object.\n",
    "        oracle.import_anndata_as_raw_count(adata=adata,\n",
    "                                        cluster_column_name=\"celltype\",\n",
    "                                        embedding_name=\"X_umap\")\n",
    "\n",
    "        # -- You can load TF info dataframe with the following code.\n",
    "        oracle.import_TF_data(TF_info_matrix=base_GRN)\n",
    "\n",
    "        # -- knn imputation, this step is needed for the whole ctrl\n",
    "        # Perform PCA\n",
    "        oracle.perform_PCA()\n",
    "\n",
    "        # Select important PCs\n",
    "        plt.plot(np.cumsum(oracle.pca.explained_variance_ratio_)[:100])\n",
    "        n_comps = np.where(np.diff(np.diff(np.cumsum(oracle.pca.explained_variance_ratio_))>0.002))[0][0]\n",
    "        plt.axvline(n_comps, c=\"k\")\n",
    "        plt.show()\n",
    "        print(n_comps)\n",
    "        n_comps = min(n_comps, 50)\n",
    "\n",
    "        n_cell = oracle.adata.shape[0]\n",
    "        print(f\"cell number is :{n_cell}\")\n",
    "\n",
    "        k = int(0.025*n_cell)\n",
    "        print(f\"Auto-selected k is :{k}\")\n",
    "\n",
    "        oracle.knn_imputation(n_pca_dims=n_comps, k=k, balanced=True, b_sight=k*8,\n",
    "                            b_maxl=k*4, n_jobs=4)\n",
    "\n",
    "        # model_prefix = ''\n",
    "        os.makedirs(save_dir, exist_ok=True)\n",
    "        # -- save the oracle\n",
    "        oracle.to_hdf5(os.path.join(save_dir, \"ctrl.celloracle.oracle\"))\n",
    "\n",
    "        # -- get the links\n",
    "        # Calculate GRN for each population in \"louvain_annot\" clustering unit.\n",
    "        # This step may take some time.(~30 minutes)\n",
    "        links = oracle.get_links(cluster_name_for_GRN_unit=\"celltype\", alpha=10,\n",
    "                                verbose_level=10)\n",
    "\n",
    "        # -- Save Links object.\n",
    "        links.to_hdf5(file_path=os.path.join(save_dir, \"ctrl.celloracle.links\"))\n",
    "\n",
    "\n",
    "    # -- filter and get the coef_mtx\n",
    "    links.filter_links(threshold_number=threshold_number,\n",
    "                        p=0.001,\n",
    "                        weight='coef_abs')\n",
    "    oracle.get_cluster_specific_TFdict_from_Links(links_object=links)\n",
    "    oracle.fit_GRN_for_simulation(alpha=10,\n",
    "                                use_cluster_specific_TFdict=True)\n",
    "    \n",
    "    ###################################################\n",
    "    # - get all the TFs in the base_GRN\n",
    "    TFdict = import_TF_data(TF_info_matrix=base_GRN)\n",
    "    tf_target_dict = {}\n",
    "    for target, gene_set in TFdict.items():\n",
    "        for tf in gene_set:\n",
    "            if tf not in tf_target_dict:\n",
    "                tf_target_dict[tf] = []\n",
    "                tf_target_dict[tf].append(target)\n",
    "            else:\n",
    "                tf_target_dict[tf].append(target)\n",
    "    total_tf_list = list(tf_target_dict.keys())\n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "    #####################################################\n",
    "    var_names = list(adata.var_names)    \n",
    "    single_total_perts = np.intersect1d(total_tf_list, adata.var_names)\n",
    "    common_perts = np.intersect1d(single_total_perts, L1000_total_perts)\n",
    "    print('L1000_total_perts num: ', len(L1000_total_perts))\n",
    "    print('common_perts num: ', len(common_perts))\n",
    "    print('common var to L1000 data is: ', len(np.intersect1d(var_names, adata_L1000.var_names)))\n",
    "\n",
    "    break\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "L1000_total_perts num:  3620\n",
      "common_perts num:  306\n"
     ]
    }
   ],
   "source": [
    "single_total_perts = np.intersect1d(total_tf_list, adata.var_names)\n",
    "common_perts = np.intersect1d(single_total_perts, L1000_total_perts)\n",
    "print('L1000_total_perts num: ', len(L1000_total_perts))\n",
    "print('common_perts num: ', len(common_perts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "###########################################\n",
    "celltype = adata.obs['celltype'].unique()[0]\n",
    "\n",
    "# - get the tf_GRN_dict, to check whether pert have regulatory relations\n",
    "gene_GRN_mtx = oracle.coef_matrix_per_cluster[celltype].copy()\n",
    "tf_GRN_mtx = gene_GRN_mtx[~(gene_GRN_mtx == 0).all(axis=1)]\n",
    "# - get TF-target pair and the regulatory values\n",
    "tf_GRN_dict = {} # the tf to targets\n",
    "for i in range(len(tf_GRN_mtx)):\n",
    "    tmp = tf_GRN_mtx.iloc[i,:]\n",
    "    tmp = tmp[tmp!=0]\n",
    "\n",
    "    tf_GRN_dict[tf_GRN_mtx.index[i]] = {}\n",
    "    for j in range(len(tmp)):\n",
    "        tf_GRN_dict[tf_GRN_mtx.index[i]][tmp.index[j]] = tmp.values[j]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5155 genes were found in the adata. Note that Celloracle is intended to use around 1000-3000 genes, so the behavior with this number of genes may differ from what is expected.\n",
      "WARNING: adata.X seems to be already log-transformed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'celltype' as categorical\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# - get oracle_ctrl\n",
    "adata_rna.obs['celltype'] = cell_line_bulk\n",
    "adata_ctrl = adata_rna.copy()\n",
    "# keep raw cont data before log transformation\n",
    "adata_ctrl.raw = adata_ctrl\n",
    "\n",
    "# the result will be recovered in normalized_count\n",
    "if not isinstance(adata_ctrl.raw.X, np.ndarray):\n",
    "    adata_ctrl.layers[\"raw_count\"] = (np.exp(adata_ctrl.raw.X.toarray())-1).copy()\n",
    "else:\n",
    "    adata_ctrl.layers[\"raw_count\"] = (np.exp(adata_ctrl.raw.X)-1).copy()\n",
    "    \n",
    "sc.pp.scale(adata_ctrl)\n",
    "# PCA\n",
    "sc.tl.pca(adata_ctrl, svd_solver='arpack', random_state=2022)\n",
    "\n",
    "# Diffusion map\n",
    "sc.pp.neighbors(adata_ctrl, n_neighbors=4, n_pcs=20, random_state=2022)\n",
    "sc.tl.umap(adata_ctrl,random_state=2022)\n",
    "\n",
    "# Instantiate Oracle object\n",
    "oracle_ctrl = co.Oracle()\n",
    "\n",
    "# In this notebook, we use the unscaled mRNA count for the nput of Oracle object.\n",
    "adata_ctrl.X = adata_ctrl.layers[\"raw_count\"].copy()\n",
    "\n",
    "# Instantiate Oracle object.\n",
    "oracle_ctrl.import_anndata_as_raw_count(adata=adata_ctrl,\n",
    "                                cluster_column_name=\"celltype\",\n",
    "                                embedding_name=\"X_umap\")\n",
    "\n",
    "# You can load TF info dataframe with the following code.\n",
    "oracle_ctrl.import_TF_data(TF_info_matrix=base_GRN)\n",
    "\n",
    "# get the imputed_count, here we dont do the impute to get the prediction\n",
    "oracle_ctrl.adata.layers[\"imputed_count\"] = oracle_ctrl.adata.layers[\"normalized_count\"].copy()\n",
    "\n",
    "# get the coef from the whole ctrl\n",
    "oracle_ctrl.coef_matrix_per_cluster = oracle.coef_matrix_per_cluster\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|▍         | 12/306 [10:43<4:18:22, 52.73s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARX is not in the tf_GRN_dict, no targets\n",
      "ARX is filtered\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  9%|▉         | 27/306 [23:01<4:03:47, 52.43s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CBFB is not in the tf_GRN_dict, no targets\n",
      "CBFB is filtered\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 11%|█▏        | 35/306 [29:08<3:48:35, 50.61s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CREB3L1 is not in the tf_GRN_dict, no targets\n",
      "CREB3L1 is filtered\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 13%|█▎        | 41/306 [33:30<3:34:07, 48.48s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DLX2 is not in the tf_GRN_dict, no targets\n",
      "DLX2 is filtered\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 17%|█▋        | 53/306 [43:09<3:41:42, 52.58s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ELF4 is not in the tf_GRN_dict, no targets\n",
      "ELF4 is filtered\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 27%|██▋       | 84/306 [1:09:24<3:13:22, 52.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GLI3 is not in the tf_GRN_dict, no targets\n",
      "GLI3 is filtered\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 30%|██▉       | 91/306 [1:14:38<2:58:39, 49.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HERPUD1 is not in the tf_GRN_dict, no targets\n",
      "HERPUD1 is filtered\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 32%|███▏      | 97/306 [1:19:15<2:50:46, 49.03s/it]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_447955/4102785368.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m     21\u001b[0m         \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m     \u001b[0;31m# Enter perturbation conditions to simulate signal propagation after the perturbation.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m     oracle_ctrl.simulate_shift(perturb_condition=goi_dict,\n\u001b[0m\u001b[1;32m     26\u001b[0m                         n_propagation=3)\n\u001b[1;32m     27\u001b[0m     \u001b[0;31m# - get the prediction; delta_X = simulated_count - imputed_count\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m     \u001b[0mdelta_X\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msimulated_count\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0moracle_ctrl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"delta_X\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moracle_ctrl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"simulated_count\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data1/lichen/anaconda3/envs/scGPT/lib/python3.9/site-packages/celloracle/trajectory/oracle_core.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, perturb_condition, GRN_unit, n_propagation, ignore_warning, use_randomized_GRN, clip_delta_X)\u001b[0m\n\u001b[1;32m    690\u001b[0m                 \u001b[0mHowever\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ma\u001b[0m \u001b[0mhigher\u001b[0m \u001b[0mnumber\u001b[0m \u001b[0mof\u001b[0m \u001b[0mpropagation\u001b[0m \u001b[0mmay\u001b[0m \u001b[0mcause\u001b[0m \u001b[0mmore\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0mnoise\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    691\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    692\u001b[0m             \u001b[0mclip_delta_X\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mbool\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIf\u001b[0m \u001b[0msimulated\u001b[0m \u001b[0mgene\u001b[0m \u001b[0mexpression\u001b[0m \u001b[0mshift\u001b[0m \u001b[0mcan\u001b[0m \u001b[0mlead\u001b[0m \u001b[0mto\u001b[0m \u001b[0mgene\u001b[0m \u001b[0mexpression\u001b[0m \u001b[0mvalue\u001b[0m \u001b[0mthat\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0moutside\u001b[0m \u001b[0mof\u001b[0m \u001b[0mWT\u001b[0m \u001b[0mdistribution\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msuch\u001b[0m \u001b[0mgene\u001b[0m \u001b[0mexpression\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mclipped\u001b[0m \u001b[0mto\u001b[0m \u001b[0mWT\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    693\u001b[0m         \"\"\"\n\u001b[0;32m--> 694\u001b[0;31m         self.__simulate_shift(perturb_condition=perturb_condition,\n\u001b[0m\u001b[1;32m    695\u001b[0m                                       \u001b[0mGRN_unit\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mGRN_unit\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    696\u001b[0m                                       \u001b[0mn_propagation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn_propagation\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    697\u001b[0m                                       \u001b[0mignore_warning\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mignore_warning\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data1/lichen/anaconda3/envs/scGPT/lib/python3.9/site-packages/celloracle/trajectory/oracle_core.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, perturb_condition, GRN_unit, n_propagation, ignore_warning, use_randomized_GRN, n_min, n_max, clip_delta_X)\u001b[0m\n\u001b[1;32m    848\u001b[0m                 \u001b[0mcells_in_the_cluster_bool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mcluster_info\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mcluster\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    849\u001b[0m                 \u001b[0msimulation_input_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msimulation_input\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcells_in_the_cluster_bool\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    850\u001b[0m                 \u001b[0mgem_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgem_imputed\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcells_in_the_cluster_bool\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    851\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 852\u001b[0;31m                 simulated_in_the_cluster = _do_simulation(\n\u001b[0m\u001b[1;32m    853\u001b[0m                                              \u001b[0mcoef_matrix\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcoef_matrix\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    854\u001b[0m                                              \u001b[0msimulation_input\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msimulation_input_\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    855\u001b[0m                                              \u001b[0mgem\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgem_\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data1/lichen/anaconda3/envs/scGPT/lib/python3.9/site-packages/celloracle/trajectory/oracle_GRN.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(coef_matrix, simulation_input, gem, n_propagation)\u001b[0m\n\u001b[1;32m     38\u001b[0m     \u001b[0mdelta_input\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msimulation_input\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mgem\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m     \u001b[0mdelta_simulated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdelta_input\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_propagation\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 42\u001b[0;31m         \u001b[0mdelta_simulated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdelta_simulated\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcoef_matrix\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     43\u001b[0m         \u001b[0mdelta_simulated\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdelta_input\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdelta_input\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m         \u001b[0;31m# gene expression cannot be negative. adjust delta values to make sure that gene expression are not netavive values.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data1/lichen/anaconda3/envs/scGPT/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m   1605\u001b[0m                 )\n\u001b[1;32m   1606\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1607\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1608\u001b[0m             return self._constructor(\n\u001b[0;32m-> 1609\u001b[0;31m                 \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlvals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrvals\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mleft\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1610\u001b[0m             )\n\u001b[1;32m   1611\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1612\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_constructor_sliced\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlvals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrvals\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mleft\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "\n",
    "for pert_suffix in tqdm(common_perts):\n",
    "\n",
    "\n",
    "    # - this is for crispra\n",
    "    gois = [pert_suffix]\n",
    "    goi_dict = {}\n",
    "\n",
    "    # - all data in L1000 is knockdown\n",
    "    for goi in gois:\n",
    "        # -- if original value is zero\n",
    "        if np.mean(adata_rna[:,goi].X.toarray())==0:\n",
    "            print(f'{goi} ctrl expression is 0')\n",
    "            continue\n",
    "        # -- if the TF has no targets\n",
    "        if goi not in list(tf_GRN_dict.keys()):\n",
    "            print(f'{goi} is not in the tf_GRN_dict, no targets')\n",
    "            continue\n",
    "        goi_dict[goi] = 0\n",
    "    if len(goi_dict) == 0:\n",
    "        print(f'{pert_suffix} is filtered')\n",
    "        continue\n",
    "\n",
    "\n",
    "    # Enter perturbation conditions to simulate signal propagation after the perturbation.\n",
    "    oracle_ctrl.simulate_shift(perturb_condition=goi_dict,\n",
    "                        n_propagation=3)\n",
    "    # - get the prediction; delta_X = simulated_count - imputed_count\n",
    "    delta_X, simulated_count = oracle_ctrl.adata.layers[\"delta_X\"], oracle_ctrl.adata.layers[\"simulated_count\"]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(500, 5155)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simulated_count.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "pert = pert_suffix\n",
    "pert_gene_rank_dict = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'celltype' as categorical\n",
      "... storing 'batch' as categorical\n"
     ]
    }
   ],
   "source": [
    "# - create adata_pert\n",
    "adata_pert = adata_rna.copy()\n",
    "adata_pert.X = simulated_count\n",
    "adata_pert.X[adata_pert.X < 0] = 0\n",
    "adata_pert.obs_names = [i+f'_{pert}' for i in adata_pert.obs_names]\n",
    "\n",
    "# - adata_ctrl\n",
    "adata_ctrl = adata_rna.copy()\n",
    "\n",
    "adata_ctrl.obs['batch'] = 'ctrl'\n",
    "adata_pert.obs['batch'] = 'pert'\n",
    "\n",
    "adata_concat = ad.concat([adata_ctrl, adata_pert])\n",
    "\n",
    "# - cal de genes\n",
    "rankby_abs = False\n",
    "\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata_concat,\n",
    "    groupby='batch',\n",
    "    reference='ctrl',\n",
    "    rankby_abs=rankby_abs,\n",
    "    n_genes=len(adata_concat.var),\n",
    "    use_raw=False,\n",
    "    method = 'wilcoxon'\n",
    ")\n",
    "de_genes = pd.DataFrame(adata_concat.uns['rank_genes_groups']['names'])\n",
    "pvals = pd.DataFrame(adata_concat.uns['rank_genes_groups']['pvals'])\n",
    "pvals_adj = pd.DataFrame(adata_concat.uns['rank_genes_groups']['pvals_adj'])\n",
    "scores = pd.DataFrame(adata_concat.uns['rank_genes_groups']['scores'])\n",
    "logfoldchanges = pd.DataFrame(adata_concat.uns['rank_genes_groups']['logfoldchanges'])\n",
    "\n",
    "# - get gene_score\n",
    "gene_score = pd.DataFrame({'gene':list(de_genes['pert']),\n",
    "                            'z-score':list(scores['pert'])})\n",
    "\n",
    "pert_gene_rank_dict[pert] = (list(de_genes['pert']), list(scores['pert']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 对所有cell line运行统一流程"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 36720 × 978\n",
       "    obs: 'sig_id', 'pert_id', 'pert_iname', 'pert_type', 'cell_id', 'pert_dose', 'pert_dose_unit', 'pert_idose', 'pert_time', 'pert_time_unit', 'pert_itime', 'distil_id'\n",
       "    var: 'pr_gene_id', 'pr_gene_symbol', 'pr_gene_title', 'pr_is_lm', 'pr_is_bing'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# - get cell line name\n",
    "common_cell_line = \\\n",
    "{   'A549': 'A549',\n",
    "    'HEPG2': 'HepG2',\n",
    "    'HT29': 'HT29',\n",
    "    'MCF7': 'MCF7',\n",
    "    # 'SKBR3': 'SK-BR-3',\n",
    "    'SW480': 'SW480',\n",
    "    'PC3': 'PC3',\n",
    "    'A375': 'A375',\n",
    "} # L1000 cell line : single-cell cell line\n",
    "\n",
    "# - read adata_L1000, this is processed data\n",
    "adata_L1000 = sc.read('/nfs/public/lichen/results/single_cell_perturbation/perturbation_benchmark/benchmark_data/L1000/GSE92742/adata_gene_pert.h5ad')\n",
    "adata_L1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================== cell line is A549\n",
      "adata.shape is:  (500, 5155)\n",
      "Loading prebuilt promoter base-GRN. Version: hg19_gimmemotifsv5_fpr2\n",
      "base_GRN.shape:  (37003, 1096)\n",
      "file exists\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9d6c7c6da3fa49d78c7a37d5c5372c7b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "L1000_total_perts num:  3620\n",
      "common_perts num:  306\n",
      "common var to L1000 data is:  933\n",
      "5155 genes were found in the adata. Note that Celloracle is intended to use around 1000-3000 genes, so the behavior with this number of genes may differ from what is expected.\n",
      "WARNING: adata.X seems to be already log-transformed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'celltype' as categorical\n",
      "  0%|          | 0/306 [00:00<?, ?it/s]... storing 'celltype' as categorical\n",
      "... storing 'batch' as categorical\n",
      "  0%|          | 0/306 [01:26<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "from sklearn.metrics import precision_recall_curve, auc\n",
    "\n",
    "for cell_line_bulk in list(common_cell_line.keys())[:]:\n",
    "    cell_line_single = common_cell_line[cell_line_bulk]\n",
    "    print('='*20, f'cell line is {cell_line_single}')\n",
    "    \n",
    "    #####################################################\n",
    "    \n",
    "    #===================prepare data\n",
    "    if cell_line_bulk in ['PC3', 'A375']:\n",
    "        save_dir_adata = '/nfs/public/lichen/results/single_cell_perturbation/perturbation_benchmark/benchmark_data/L1000/single_cell_data/SCP542'\n",
    "    else:\n",
    "        save_dir_adata = '/nfs/public/lichen/results/single_cell_perturbation/perturbation_benchmark/benchmark_data/L1000/single_cell_data/CNP0003658'\n",
    "    adata_rna = sc.read(os.path.join(save_dir_adata, cell_line_bulk, f'adata_{cell_line_bulk}.h5ad'))\n",
    "\n",
    "    # - consctrut corr mtx\n",
    "    if not isinstance(adata_rna.X, np.ndarray):\n",
    "        adata_rna.X = adata_rna.X.toarray()\n",
    "    # corr_mtx = np.corrcoef(adata_rna.X.T)\n",
    "    \n",
    "    # - get var_names\n",
    "    var_names = list(adata_rna.var_names)\n",
    "    \n",
    "    # - get common pert\n",
    "    adata_L1000_sub = adata_L1000[adata_L1000.obs['cell_id']==cell_line_bulk]\n",
    "    L1000_total_perts = np.unique(adata_L1000_sub.obs['pert_iname'])\n",
    "    \n",
    "    \n",
    "    n_cells_downsample = 10000\n",
    "    threshold_number = 10000\n",
    "    \n",
    "    ##########################################################\n",
    "    \n",
    "    # - get control adata\n",
    "    adata = adata_rna.copy()\n",
    "    adata.obs['celltype'] = cell_line_bulk\n",
    "    print(f'adata.shape is: ',adata.shape)\n",
    "\n",
    "    # -- get the baseGRN\n",
    "    # Load TF info which was made from mouse cell atlas dataset.\n",
    "    base_GRN = co.data.load_human_promoter_base_GRN()\n",
    "    print('base_GRN.shape: ', base_GRN.shape)\n",
    "\n",
    "    tmp_dir = '/nfs/public/lichen/results/single_cell_perturbation/perturbation_benchmark/L1000'\n",
    "    save_prefix = f'CellOracle/{cell_line_bulk}' # use result of K562 to do the direct transfer\n",
    "    os.makedirs(os.path.join(tmp_dir, save_prefix), exist_ok=True)\n",
    "\n",
    "    save_dir = os.path.join(tmp_dir, save_prefix)\n",
    "    if os.path.exists(os.path.join(save_dir, \"ctrl.celloracle.oracle\")):\n",
    "        print('file exists')\n",
    "        oracle = co.load_hdf5(os.path.join(save_dir, \"ctrl.celloracle.oracle\"))\n",
    "        links = co.load_hdf5(file_path=os.path.join(save_dir, \"ctrl.celloracle.links\"))\n",
    "        \n",
    "    else:\n",
    "\n",
    "        # - start CellOracle process for the whole ctrl\n",
    "\n",
    "        # -- keep raw cont data before log transformation\n",
    "        adata.raw = adata\n",
    "        if not isinstance(adata.raw.X, np.ndarray):\n",
    "            adata.layers[\"raw_count\"] = (np.exp(adata.raw.X.toarray())-1).copy()\n",
    "        else:\n",
    "            adata.layers[\"raw_count\"] = (np.exp(adata.raw.X)-1).copy()\n",
    "            \n",
    "        # -- get umap \n",
    "        sc.pp.scale(adata)\n",
    "        # PCA\n",
    "        sc.tl.pca(adata, svd_solver='arpack', random_state=2022)\n",
    "        # UMAP\n",
    "        sc.pp.neighbors(adata, n_neighbors=4, n_pcs=20, random_state=2022)\n",
    "        sc.tl.umap(adata,random_state=2022)\n",
    "\n",
    "        # -- Random downsampling into 30K cells if the anndata object include more than 30 K cells.\n",
    "        if adata.shape[0] > n_cells_downsample:\n",
    "            # Let's dowmsample into 30K cells\n",
    "            sc.pp.subsample(adata, n_obs=n_cells_downsample, random_state=123)\n",
    "        print(f\"Cell number is :{adata.shape[0]}\")\n",
    "\n",
    "        # -- Instantiate Oracle object\n",
    "        oracle = co.Oracle()\n",
    "\n",
    "        # -- Check data in anndata\n",
    "        print(\"Metadata columns :\", list(adata.obs.columns))\n",
    "        print(\"Dimensional reduction: \", list(adata.obsm.keys()))\n",
    "\n",
    "        # -- In this notebook, we use the unscaled mRNA count for the nput of Oracle object.\n",
    "        adata.X = adata.layers[\"raw_count\"].copy()\n",
    "\n",
    "        # -- Instantiate Oracle object.\n",
    "        oracle.import_anndata_as_raw_count(adata=adata,\n",
    "                                        cluster_column_name=\"celltype\",\n",
    "                                        embedding_name=\"X_umap\")\n",
    "\n",
    "        # -- You can load TF info dataframe with the following code.\n",
    "        oracle.import_TF_data(TF_info_matrix=base_GRN)\n",
    "\n",
    "        # -- knn imputation, this step is needed for the whole ctrl\n",
    "        # Perform PCA\n",
    "        oracle.perform_PCA()\n",
    "\n",
    "        # Select important PCs\n",
    "        plt.plot(np.cumsum(oracle.pca.explained_variance_ratio_)[:100])\n",
    "        n_comps = np.where(np.diff(np.diff(np.cumsum(oracle.pca.explained_variance_ratio_))>0.002))[0][0]\n",
    "        plt.axvline(n_comps, c=\"k\")\n",
    "        plt.show()\n",
    "        print(n_comps)\n",
    "        n_comps = min(n_comps, 50)\n",
    "\n",
    "        n_cell = oracle.adata.shape[0]\n",
    "        print(f\"cell number is :{n_cell}\")\n",
    "\n",
    "        k = int(0.025*n_cell)\n",
    "        print(f\"Auto-selected k is :{k}\")\n",
    "\n",
    "        oracle.knn_imputation(n_pca_dims=n_comps, k=k, balanced=True, b_sight=k*8,\n",
    "                            b_maxl=k*4, n_jobs=4)\n",
    "\n",
    "        # model_prefix = ''\n",
    "        os.makedirs(save_dir, exist_ok=True)\n",
    "        # -- save the oracle\n",
    "        oracle.to_hdf5(os.path.join(save_dir, \"ctrl.celloracle.oracle\"))\n",
    "\n",
    "        # -- get the links\n",
    "        # Calculate GRN for each population in \"louvain_annot\" clustering unit.\n",
    "        # This step may take some time.(~30 minutes)\n",
    "        links = oracle.get_links(cluster_name_for_GRN_unit=\"celltype\", alpha=10,\n",
    "                                verbose_level=10)\n",
    "\n",
    "        # -- Save Links object.\n",
    "        links.to_hdf5(file_path=os.path.join(save_dir, \"ctrl.celloracle.links\"))\n",
    "\n",
    "\n",
    "    # -- filter and get the coef_mtx\n",
    "    links.filter_links(threshold_number=threshold_number,\n",
    "                        p=0.001,\n",
    "                        weight='coef_abs')\n",
    "    oracle.get_cluster_specific_TFdict_from_Links(links_object=links)\n",
    "    oracle.fit_GRN_for_simulation(alpha=10,\n",
    "                                use_cluster_specific_TFdict=True)\n",
    "    \n",
    "    ###################################################\n",
    "    # - get all the TFs in the base_GRN\n",
    "    TFdict = import_TF_data(TF_info_matrix=base_GRN)\n",
    "    tf_target_dict = {}\n",
    "    for target, gene_set in TFdict.items():\n",
    "        for tf in gene_set:\n",
    "            if tf not in tf_target_dict:\n",
    "                tf_target_dict[tf] = []\n",
    "                tf_target_dict[tf].append(target)\n",
    "            else:\n",
    "                tf_target_dict[tf].append(target)\n",
    "    total_tf_list = list(tf_target_dict.keys())\n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "    #####################################################\n",
    "    var_names = list(adata.var_names)    \n",
    "    single_total_perts = np.intersect1d(total_tf_list, adata.var_names)\n",
    "    common_perts = np.intersect1d(single_total_perts, L1000_total_perts)\n",
    "    print('L1000_total_perts num: ', len(L1000_total_perts))\n",
    "    print('common_perts num: ', len(common_perts))\n",
    "    print('common var to L1000 data is: ', len(np.intersect1d(var_names, adata_L1000.var_names)))\n",
    "\n",
    "    ###########################################\n",
    "    celltype = adata.obs['celltype'].unique()[0]\n",
    "\n",
    "    # - get the tf_GRN_dict, to check whether pert have regulatory relations\n",
    "    gene_GRN_mtx = oracle.coef_matrix_per_cluster[celltype].copy()\n",
    "    tf_GRN_mtx = gene_GRN_mtx[~(gene_GRN_mtx == 0).all(axis=1)]\n",
    "    # - get TF-target pair and the regulatory values\n",
    "    tf_GRN_dict = {} # the tf to targets\n",
    "    for i in range(len(tf_GRN_mtx)):\n",
    "        tmp = tf_GRN_mtx.iloc[i,:]\n",
    "        tmp = tmp[tmp!=0]\n",
    "\n",
    "        tf_GRN_dict[tf_GRN_mtx.index[i]] = {}\n",
    "        for j in range(len(tmp)):\n",
    "            tf_GRN_dict[tf_GRN_mtx.index[i]][tmp.index[j]] = tmp.values[j]\n",
    "\n",
    "    ###########################################        \n",
    "    # - get oracle_ctrl\n",
    "    adata_rna.obs['celltype'] = cell_line_bulk\n",
    "    adata_ctrl = adata_rna.copy()\n",
    "    # keep raw cont data before log transformation\n",
    "    adata_ctrl.raw = adata_ctrl\n",
    "\n",
    "    # the result will be recovered in normalized_count\n",
    "    if not isinstance(adata_ctrl.raw.X, np.ndarray):\n",
    "        adata_ctrl.layers[\"raw_count\"] = (np.exp(adata_ctrl.raw.X.toarray())-1).copy()\n",
    "    else:\n",
    "        adata_ctrl.layers[\"raw_count\"] = (np.exp(adata_ctrl.raw.X)-1).copy()\n",
    "        \n",
    "    sc.pp.scale(adata_ctrl)\n",
    "    # PCA\n",
    "    sc.tl.pca(adata_ctrl, svd_solver='arpack', random_state=2022)\n",
    "\n",
    "    # Diffusion map\n",
    "    sc.pp.neighbors(adata_ctrl, n_neighbors=4, n_pcs=20, random_state=2022)\n",
    "    sc.tl.umap(adata_ctrl,random_state=2022)\n",
    "\n",
    "    # Instantiate Oracle object\n",
    "    oracle_ctrl = co.Oracle()\n",
    "\n",
    "    # In this notebook, we use the unscaled mRNA count for the nput of Oracle object.\n",
    "    adata_ctrl.X = adata_ctrl.layers[\"raw_count\"].copy()\n",
    "\n",
    "    # Instantiate Oracle object.\n",
    "    oracle_ctrl.import_anndata_as_raw_count(adata=adata_ctrl,\n",
    "                                    cluster_column_name=\"celltype\",\n",
    "                                    embedding_name=\"X_umap\")\n",
    "\n",
    "    # You can load TF info dataframe with the following code.\n",
    "    oracle_ctrl.import_TF_data(TF_info_matrix=base_GRN)\n",
    "\n",
    "    # get the imputed_count, here we dont do the impute to get the prediction\n",
    "    oracle_ctrl.adata.layers[\"imputed_count\"] = oracle_ctrl.adata.layers[\"normalized_count\"].copy()\n",
    "\n",
    "    # get the coef from the whole ctrl\n",
    "    oracle_ctrl.coef_matrix_per_cluster = oracle.coef_matrix_per_cluster\n",
    "    \n",
    "    pert_gene_rank_dict = {} \n",
    "    for pert in tqdm(common_perts):\n",
    "        \n",
    "\n",
    "        # - this is for crispra\n",
    "        gois = [pert]\n",
    "        goi_dict = {}\n",
    "\n",
    "        # - all data in L1000 is knockdown\n",
    "        for goi in gois:\n",
    "            # -- if original value is zero\n",
    "            if np.mean(adata_rna[:,goi].X.toarray())==0:\n",
    "                print(f'{goi} ctrl expression is 0')\n",
    "                continue\n",
    "            # -- if the TF has no targets\n",
    "            if goi not in list(tf_GRN_dict.keys()):\n",
    "                print(f'{goi} is not in the tf_GRN_dict, no targets')\n",
    "                continue\n",
    "            goi_dict[goi] = 0\n",
    "        if len(goi_dict) == 0:\n",
    "            print(f'{pert} is filtered')\n",
    "            continue\n",
    "\n",
    "\n",
    "        # Enter perturbation conditions to simulate signal propagation after the perturbation.\n",
    "        oracle_ctrl.simulate_shift(perturb_condition=goi_dict,\n",
    "                            n_propagation=3)\n",
    "        # - get the prediction; delta_X = simulated_count - imputed_count\n",
    "        delta_X, simulated_count = oracle_ctrl.adata.layers[\"delta_X\"], oracle_ctrl.adata.layers[\"simulated_count\"]\n",
    "\n",
    "\n",
    "        # - create adata_pert\n",
    "        adata_pert = adata_rna.copy()\n",
    "        adata_pert.X = simulated_count\n",
    "        adata_pert.X[adata_pert.X < 0] = 0\n",
    "        adata_pert.obs_names = [i+f'_{pert}' for i in adata_pert.obs_names]\n",
    "\n",
    "        # - adata_ctrl\n",
    "        adata_ctrl = adata_rna.copy()\n",
    "\n",
    "        adata_ctrl.obs['batch'] = 'ctrl'\n",
    "        adata_pert.obs['batch'] = 'pert'\n",
    "\n",
    "        adata_concat = ad.concat([adata_ctrl, adata_pert])\n",
    "\n",
    "        # - cal de genes\n",
    "        rankby_abs = False\n",
    "\n",
    "        sc.tl.rank_genes_groups(\n",
    "            adata_concat,\n",
    "            groupby='batch',\n",
    "            reference='ctrl',\n",
    "            rankby_abs=rankby_abs,\n",
    "            n_genes=len(adata_concat.var),\n",
    "            use_raw=False,\n",
    "            method = 'wilcoxon'\n",
    "        )\n",
    "        de_genes = pd.DataFrame(adata_concat.uns['rank_genes_groups']['names'])\n",
    "        pvals = pd.DataFrame(adata_concat.uns['rank_genes_groups']['pvals'])\n",
    "        pvals_adj = pd.DataFrame(adata_concat.uns['rank_genes_groups']['pvals_adj'])\n",
    "        scores = pd.DataFrame(adata_concat.uns['rank_genes_groups']['scores'])\n",
    "        logfoldchanges = pd.DataFrame(adata_concat.uns['rank_genes_groups']['logfoldchanges'])\n",
    "\n",
    "        # - get gene_score\n",
    "        gene_score = pd.DataFrame({'gene':list(de_genes['pert']),\n",
    "                                    'z-score':list(scores['pert'])})\n",
    "\n",
    "        pert_gene_rank_dict[pert] = (list(de_genes['pert']), list(scores['pert']))\n",
    "        \n",
    "        break\n",
    "        \n",
    "    # save_dir = '/nfs/public/lichen/results/single_cell_perturbation/perturbation_benchmark/L1000'\n",
    "    # save_prefix = f'CellOracle/{cell_line_bulk}' # use result of K562 to do the direct transfer\n",
    "    # os.makedirs(os.path.join(save_dir, save_prefix), exist_ok=True)\n",
    "\n",
    "    # import json\n",
    "    # with open(os.path.join(save_dir, save_prefix, 'pert_gene_rank_dict.json'), 'w') as f:\n",
    "    #     json.dump(pert_gene_rank_dict, f)\n",
    "        \n",
    "\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_1 = adata_rna.X + delta_X\n",
    "tmp_2 = simulated_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({True: 2244235, False: 333265})\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "print(Counter(list((tmp_1 == tmp_2).ravel())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_1 = adata_rna.X\n",
    "tmp_2 = simulated_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({True: 2244159, False: 333341})\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "print(Counter(list((tmp_1 == tmp_2).ravel())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_1 = adata_rna.X\n",
    "tmp_2 = oracle_ctrl.adata.layers[\"imputed_count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({True: 2244235, False: 333265})\n"
     ]
    }
   ],
   "source": [
    "print(Counter(list((tmp_1 == tmp_2).ravel())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'celltype' as categorical\n",
      "... storing 'batch' as categorical\n"
     ]
    }
   ],
   "source": [
    "# - create adata_pert\n",
    "adata_pert = adata_rna.copy()\n",
    "adata_pert.X = simulated_count\n",
    "# adata_pert.X += delta_X\n",
    "adata_pert.X[adata_pert.X < 0] = 0\n",
    "adata_pert.obs_names = [i+f'_{pert}' for i in adata_pert.obs_names]\n",
    "\n",
    "# - adata_ctrl\n",
    "adata_ctrl = adata_rna.copy()\n",
    "\n",
    "adata_ctrl.obs['batch'] = 'ctrl'\n",
    "adata_pert.obs['batch'] = 'pert'\n",
    "\n",
    "adata_concat = ad.concat([adata_ctrl, adata_pert])\n",
    "\n",
    "# - cal de genes\n",
    "rankby_abs = False\n",
    "\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata_concat,\n",
    "    groupby='batch',\n",
    "    reference='ctrl',\n",
    "    rankby_abs=rankby_abs,\n",
    "    n_genes=len(adata_concat.var),\n",
    "    use_raw=False,\n",
    "    method = 'wilcoxon'\n",
    ")\n",
    "de_genes = pd.DataFrame(adata_concat.uns['rank_genes_groups']['names'])\n",
    "pvals = pd.DataFrame(adata_concat.uns['rank_genes_groups']['pvals'])\n",
    "pvals_adj = pd.DataFrame(adata_concat.uns['rank_genes_groups']['pvals_adj'])\n",
    "scores = pd.DataFrame(adata_concat.uns['rank_genes_groups']['scores'])\n",
    "logfoldchanges = pd.DataFrame(adata_concat.uns['rank_genes_groups']['logfoldchanges'])\n",
    "\n",
    "# - get gene_score\n",
    "gene_score = pd.DataFrame({'gene':list(de_genes['pert']),\n",
    "                            'z-score':list(scores['pert'])})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>z-score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>INSL4</td>\n",
       "      <td>0.212520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLRT2</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FOXJ2</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FOXJ3</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FOXL2NB</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5150</th>\n",
       "      <td>POLE3</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5151</th>\n",
       "      <td>PPIA</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5152</th>\n",
       "      <td>RALB</td>\n",
       "      <td>-0.062081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5153</th>\n",
       "      <td>AURKA</td>\n",
       "      <td>-0.087154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5154</th>\n",
       "      <td>AFP</td>\n",
       "      <td>-1.076285</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5155 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene   z-score\n",
       "0       INSL4  0.212520\n",
       "1       FLRT2  0.000000\n",
       "2       FOXJ2  0.000000\n",
       "3       FOXJ3  0.000000\n",
       "4     FOXL2NB  0.000000\n",
       "...       ...       ...\n",
       "5150    POLE3  0.000000\n",
       "5151     PPIA  0.000000\n",
       "5152     RALB -0.062081\n",
       "5153    AURKA -0.087154\n",
       "5154      AFP -1.076285\n",
       "\n",
       "[5155 rows x 2 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>z-score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>INSL4</td>\n",
       "      <td>0.210220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRPRB</td>\n",
       "      <td>0.002847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAD2L1BP</td>\n",
       "      <td>0.002518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UFL1</td>\n",
       "      <td>0.002518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NEK2</td>\n",
       "      <td>0.002518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5150</th>\n",
       "      <td>MCL1</td>\n",
       "      <td>-0.003942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5151</th>\n",
       "      <td>MAT2B</td>\n",
       "      <td>-0.004270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5152</th>\n",
       "      <td>RALB</td>\n",
       "      <td>-0.062300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5153</th>\n",
       "      <td>AURKA</td>\n",
       "      <td>-0.087482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5154</th>\n",
       "      <td>AFP</td>\n",
       "      <td>-1.642347</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5155 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene   z-score\n",
       "0        INSL4  0.210220\n",
       "1        SRPRB  0.002847\n",
       "2     MAD2L1BP  0.002518\n",
       "3         UFL1  0.002518\n",
       "4         NEK2  0.002518\n",
       "...        ...       ...\n",
       "5150      MCL1 -0.003942\n",
       "5151     MAT2B -0.004270\n",
       "5152      RALB -0.062300\n",
       "5153     AURKA -0.087482\n",
       "5154       AFP -1.642347\n",
       "\n",
       "[5155 rows x 2 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AFP'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pert"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['INSL4',\n",
       " 'SRPRB',\n",
       " 'MAD2L1BP',\n",
       " 'UFL1',\n",
       " 'NEK2',\n",
       " 'FAM20B',\n",
       " 'DTX3L',\n",
       " 'PDZD8',\n",
       " 'CREB3',\n",
       " 'RFC2']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pert_gene_rank_dict[pert][0][0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.21022038161754608,\n",
       " 0.0028467343654483557,\n",
       " 0.0025182650424540043,\n",
       " 0.0025182650424540043,\n",
       " 0.0025182650424540043,\n",
       " 0.002408775268122554,\n",
       " 0.002408775268122554,\n",
       " 0.0022992854937911034,\n",
       " 0.0022992854937911034,\n",
       " 0.002189795719459653,\n",
       " 0.002080305712297559,\n",
       " 0.002080305712297559,\n",
       " 0.002080305712297559,\n",
       " 0.002080305712297559,\n",
       " 0.0019708159379661083,\n",
       " 0.0019708159379661083,\n",
       " 0.0019708159379661083,\n",
       " 0.0019708159379661083,\n",
       " 0.0019708159379661083,\n",
       " 0.0019708159379661083,\n",
       " 0.0018613262800499797,\n",
       " 0.0018613262800499797,\n",
       " 0.0018613262800499797,\n",
       " 0.0018613262800499797,\n",
       " 0.0018613262800499797,\n",
       " 0.0018613262800499797,\n",
       " 0.0017518365057185292,\n",
       " 0.0017518365057185292,\n",
       " 0.0017518365057185292,\n",
       " 0.0017518365057185292,\n",
       " 0.0017518365057185292,\n",
       " 0.0017518365057185292,\n",
       " 0.0017518365057185292,\n",
       " 0.0017518365057185292,\n",
       " 0.0017518365057185292,\n",
       " 0.0016423467313870788,\n",
       " 0.0016423467313870788,\n",
       " 0.0016423467313870788,\n",
       " 0.0016423467313870788,\n",
       " 0.0016423467313870788,\n",
       " 0.0016423467313870788,\n",
       " 0.0016423467313870788,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0015328569570556283,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0014233671827241778,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.0013138774083927274,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.001204387634061277,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0010948978597298265,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0009854079689830542,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0008759182528592646,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0007664284785278141,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0006569387041963637,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0005474489298649132,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.0004379591264296323,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00032846935209818184,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " 0.00021897956321481615,\n",
       " ...]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pert_gene_rank_dict[pert][1][0:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# np.unique(delta_X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.        , 0.        , 0.79869837, ..., 0.        , 0.        ,\n",
       "        0.79869837],\n",
       "       [0.5719357 , 0.        , 0.9334965 , ..., 0.        , 0.        ,\n",
       "        0.        ],\n",
       "       [0.        , 0.        , 0.8364108 , ..., 0.        , 0.        ,\n",
       "        0.50321716],\n",
       "       ...,\n",
       "       [0.        , 0.        , 1.5602019 , ..., 0.        , 0.        ,\n",
       "        0.        ],\n",
       "       [0.        , 0.        , 1.1712078 , ..., 0.55501235, 0.        ,\n",
       "        0.        ],\n",
       "       [0.9632272 , 0.        , 0.59336513, ..., 0.        , 0.        ,\n",
       "        0.59336513]], dtype=float32)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oracle_ctrl.adata.layers[\"imputed_count\"] + "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ True,  True,  True, ...,  True,  True,  True],\n",
       "       [ True,  True,  True, ...,  True,  True,  True],\n",
       "       [ True,  True,  True, ...,  True,  True,  True],\n",
       "       ...,\n",
       "       [ True,  True,  True, ...,  True,  True,  True],\n",
       "       [ True,  True,  True, ...,  True,  True,  True],\n",
       "       [ True,  True,  True, ...,  True,  True,  True]])"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oracle_ctrl.adata.layers[\"imputed_count\"] == simulated_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = oracle_ctrl.adata.layers[\"imputed_count\"] == simulated_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({True: 2577391, False: 109})\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "print(Counter(list(tmp.ravel())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = (oracle_ctrl.adata.layers[\"imputed_count\"] + delta_X) == simulated_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "        [0., 0., 0., ..., 0., 0., 0.],\n",
       "        [0., 0., 0., ..., 0., 0., 0.],\n",
       "        ...,\n",
       "        [0., 0., 0., ..., 0., 0., 0.],\n",
       "        [0., 0., 0., ..., 0., 0., 0.],\n",
       "        [0., 0., 0., ..., 0., 0., 0.]]),\n",
       " array([[0.        , 0.        , 0.79869837, ..., 0.        , 0.        ,\n",
       "         0.79869837],\n",
       "        [0.57193571, 0.        , 0.93349648, ..., 0.        , 0.        ,\n",
       "         0.        ],\n",
       "        [0.        , 0.        , 0.83641082, ..., 0.        , 0.        ,\n",
       "         0.50321716],\n",
       "        ...,\n",
       "        [0.        , 0.        , 1.56020188, ..., 0.        , 0.        ,\n",
       "         0.        ],\n",
       "        [0.        , 0.        , 1.17120779, ..., 0.55501235, 0.        ,\n",
       "         0.        ],\n",
       "        [0.96322721, 0.        , 0.59336513, ..., 0.        , 0.        ,\n",
       "         0.59336513]]))"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "delta_X, simulated_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.16837092994967873, 7.173549652099609)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(delta_X), np.max(simulated_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-1.2854061126708984, 0.0)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.min(delta_X), np.min(simulated_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scGPT",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
